[WIP]Improve connection error banner

kuv2707 opened the pull request:
This fixes the first two checkpoints of #33924 


> - [ ] Change the banner text to:
>  `Unable to connect to Zulip. Trying again in {x} seconds… [Try now]`
> - [ ] Show a loading indicator on the button any time that a retry is currently in progress.


Fixes part of: #33924 



<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

[CZO](https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.8E.AF.20Glitchy.20connection.20error.20popup.20banner/with/2122885)

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

kuv2707 commented on web/src/server_events.js:
```diff
@@ -201,46 +201,43 @@ function get_events({dont_block = false} = {}) {
-                }
-            } catch (error) {
-                blueslip.error("Failed to handle get_events error", undefined, error);
+            get_events_xhr = undefined;
```
This diff corresponds to removing the `try catch`

timabbott replied:
Hmm. I think of this try/catch is really important, because the consequence of this code path throwing an exception is a complete failure of the client to live-update further, which is Very Bad as a failure mode.

kuv2707 replied:
Oh, alright. It didn't seem to me that any of the line there would throw an exception.
I have restructured the code around this.

kuv2707 commented:
@alya It would be great if you could manually test this and provide some feedback. The third point in the issue description seems less worthy for the effort, since the loading indicator on the button will only ever last for a second.

alya commented:
This feels pretty good overall!

I think we should do something about the initial behavior, where it's constantly flashing between "trying now" and the countdown. Maybe @timabbott has a good suggestion.

alya commented:
Another thing I noted was that it should be "1 second", not "1 seconds". I expect this also means we're not using appropriate internationalization for strings with numerals; see https://zulip.readthedocs.io/en/latest/translating/internationalization.html#plurals-and-lists.

alya commented:
I agree that the button behavior feels fine as it's implemented, so I dropping the third point seems fine to me.

kuv2707 commented:
Fixed the internationalization of "second(s)".

timabbott commented:
I'd like to see much more careful fine-grained commits here. This is one of our most sensitive/important code paths, so refactoring that is hard to read cannot be merged to it.

kuv2707 commented on web/src/message_fetch.ts:
> I think we should do something about the initial behavior, where it's constantly flashing between "trying now" and the countdown. Maybe @timabbott has a good suggestion.

I'd need some help with this. The reason for this commit is to avoid having the countdown on the `connection error banner` appear random.
The constant initial flashing is due to the small time gaps between retries in `message_fetch`. We could show the banner after a fixed amount of failures, just like in `server_events`, but that would then cause the banner to appear late. This doesn't seem to be a viable solution.

sayamsamal replied:
We can have a min_limit for the number of seconds where we want to show the timer countdown. For short periods of time, let's just show "Retrying soon", and for times greater than, say, 5 seconds let's show the timer.

kuv2707 replied:
I was thinking the same. I'll implement it.

kuv2707 commented:
@timabbott  I've tried to restructure the commits, and made the changes less intrusive to the sensitive code paths.
The changes to the popup banner are a bit huge, since they can't be meaningfully separated.
I hope it is easier to review now.

kuv2707 commented:
@zulipbot remove "has conflicts"

sayamsamal commented:
Also, I don't see the logic prioritizing the `events` codepath banner over the `message_fetch` banner, which was discussed at CZO: [#issues > 🎯 Glitchy connection error popup banner @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.8E.AF.20Glitchy.20connection.20error.20popup.20banner/near/2122163).

sayamsamal commented on web/src/server_events.js:
```diff
@@ -239,7 +241,9 @@ function get_events({dont_block = false} = {}) {
             }
 
-            const retry_delay_secs = util.get_retry_backoff_seconds(xhr, get_events_failures);
+            if (retry_delay_secs === undefined) {
+                retry_delay_secs = util.get_retry_backoff_seconds(xhr, get_events_failures);
+            }
```
Why would `retry_delay_secs` be undefined here if we are already defining it earlier?

kuv2707 replied:
This is just one of those bugs that are prevented by just having typescript. The first argument is supposed to be `opener`, which I accidentally missed, and got no complains from eslint etc. The object we're passing is supposed to be the second arg.

I'll push the fix.

kuv2707 replied:
Fixed. I hope this module soon converts to TS.

sayamsamal commented on web/src/popup_banners.ts:
```diff
@@ -69,31 +54,100 @@ export function close_found_missing_unreads_banner(): void {
+    }
+    return $t(
+        {
+            defaultMessage:
+                "{seconds, plural, one {Unable to connect to Zulip. Trying again in {seconds} second...} other {Unable to connect to Zulip. Trying again in {seconds} seconds...}}",
```
The "Unable to connect to Zulip." part of the label is constant, so you don't have to pass it for ICU MessageFormat.

kuv2707 replied:
Sure, i'll shift it out.

kuv2707 commented:
> Also, I don't see the logic prioritizing the `events` codepath banner over the `message_fetch` banner, which was discussed at CZO: [#issues > 🎯 Glitchy connection error popup banner @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.8E.AF.20Glitchy.20connection.20error.20popup.20banner/near/2122163).

That ended up not being required after implementing the mutual exclusivity scheme. The code path that first triggered the opening of the banner can control its lifecycle, and the other will be ignored. This avoids the conflicts occuring previously.

kuv2707 commented:
@sayamsamal I have addressed the feedback. Also posted an update: [#issues > 🎯 Glitchy connection error popup banner @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.8E.AF.20Glitchy.20connection.20error.20popup.20banner/near/2125626)

timabbott commented:
Closing in favor of #34031, which just merged.
